From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 19 Feb 2026 04:15:20 +0100
Subject: xdg-permission-store: Always use the permission ID to get or set
 Apps permissions

Snaps use a permssion ID that does not match the application ID, while
the permissions store portal API refers to the app (IDs).

Now, we do not want to change this, given that portals assume that what
they receive the app-id, that is the basename of the desktop ID of the
application, so we do it here.

When an app request is received, we check the desktop file of the given
app and if that belongs to a snap application, then we use the snap
name as the permission id.

In this way we can both support the standard flatpak apps and the snap
ones, without changing the way they saved the permissions.

This implies some minimal overhead for non-snap applications, since we
are now looking up into desktop files, but it should be quite irrelevant

Potentially we could, in future, just track the files in the well known
paths in which the sandboxed applications put their desktop files, so
that we can optimize this a bit with an hashmap.

Origin: https://github.com/flatpak/xdg-desktop-portal/pull/1764
---
 document-portal/permission-db.c        | 38 ++++++++++++++++++
 document-portal/permission-db.h        |  2 +
 document-portal/xdg-permission-store.c | 12 +++---
 tests/test_permission_store.py         | 71 +++++++++++++++++++++++++++++-----
 4 files changed, 108 insertions(+), 15 deletions(-)

diff --git a/document-portal/permission-db.c b/document-portal/permission-db.c
index d3f35c2..221378b 100644
--- a/document-portal/permission-db.c
+++ b/document-portal/permission-db.c
@@ -25,6 +25,7 @@
 
 #include <string.h>
 #include <fcntl.h>
+#include <gio/gdesktopappinfo.h>
 #include <stdio.h>
 #include <stdlib.h>
 #if HAVE_SYS_STATFS_H
@@ -1253,3 +1254,40 @@ permission_db_entry_print_string (PermissionDbEntry *entry,
 {
   return g_variant_print_string ((GVariant *) entry, string, FALSE);
 }
+
+char *
+permissions_id_for_app_id (const char *app_id)
+{
+  g_autoptr (GDesktopAppInfo) app_info = NULL;
+  g_autofree char *desktop_id = NULL;
+
+  if (app_id == NULL || *app_id == '\0' || g_str_has_prefix (app_id, "snap."))
+    return g_strdup (app_id);
+
+  /* We could optimize the case for flatpak's, by something like
+   *   !g_str_has_prefix (app_id, "snap.") && strchr (app_id, '.')
+   * but snaps are going to support normal desktop IDs, and so it's better
+   * not to optimize something that we need to drop soon.
+   */
+
+  g_debug ("Looking for app ID for %s", app_id);
+  desktop_id = g_strconcat (app_id, ".desktop", NULL);
+  app_info = g_desktop_app_info_new (desktop_id);
+
+  if (!app_info)
+    return g_strdup (app_id);
+
+  if (g_desktop_app_info_has_key (app_info, "X-SnapInstanceName"))
+    {
+      g_autofree char *snap_instance_name = NULL;
+
+      g_debug ("App ID for %s is provided by snap desktop ID", desktop_id);
+
+      snap_instance_name = g_desktop_app_info_get_string (app_info,
+                                                          "X-SnapInstanceName");
+      g_return_val_if_fail (snap_instance_name != NULL, g_strdup (app_id));
+      return g_strconcat ("snap.", snap_instance_name, NULL);
+    }
+
+  return g_strdup (app_id);
+}
diff --git a/document-portal/permission-db.h b/document-portal/permission-db.h
index 52ca339..f91022f 100644
--- a/document-portal/permission-db.h
+++ b/document-portal/permission-db.h
@@ -92,6 +92,8 @@ PermissionDbEntry  *permission_db_entry_set_app_permissions (PermissionDbEntry *
 PermissionDbEntry  *permission_db_entry_remove_app_permissions (PermissionDbEntry *entry,
                                                                 const char        *app_permissions_id);
 
+char *permissions_id_for_app_id (const char *app_id);
+
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (PermissionDb, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (PermissionDbEntry, permission_db_entry_unref)
 
diff --git a/document-portal/xdg-permission-store.c b/document-portal/xdg-permission-store.c
index 1f677ca..6598180 100644
--- a/document-portal/xdg-permission-store.c
+++ b/document-portal/xdg-permission-store.c
@@ -300,7 +300,7 @@ handle_delete_permission (XdgPermissionStore     *object,
                           const char             *app)
 {
   Table *table;
-  const char *permissions_id = NULL;
+  g_autofree char *permissions_id = NULL;
   g_autoptr(PermissionDbEntry) entry = NULL;
   g_autoptr(PermissionDbEntry) new_entry = NULL;
 
@@ -317,7 +317,7 @@ handle_delete_permission (XdgPermissionStore     *object,
       return TRUE;
     }
 
-  permissions_id = app;
+  permissions_id = permissions_id_for_app_id (app);
   new_entry = permission_db_entry_remove_app_permissions (entry, permissions_id);
   permission_db_set_entry (table->db, id, new_entry);
   emit_changed (object, table_name, id, new_entry);
@@ -338,7 +338,7 @@ handle_get_permission (XdgPermissionStore     *object,
 
   g_autoptr(PermissionDbEntry) entry = NULL;
   g_autofree const char **permission = NULL;
-  const char *permissions_id;
+  g_autofree char *permissions_id = NULL;
 
   table = lookup_table (table_name, invocation);
   if (table == NULL)
@@ -353,7 +353,7 @@ handle_get_permission (XdgPermissionStore     *object,
       return TRUE;
     }
 
-  permissions_id = app;
+  permissions_id = permissions_id_for_app_id (app);
   permission = permission_db_entry_list_permissions (entry, permissions_id);
 
   xdg_permission_store_complete_get_permission (object, invocation, permission);
@@ -433,7 +433,7 @@ handle_set_permission (XdgPermissionStore     *object,
                        const gchar *const     *permissions)
 {
   Table *table;
-  const char *permissions_id;
+  g_autofree char *permissions_id = NULL;
   g_autoptr(PermissionDbEntry) entry = NULL;
   g_autoptr(PermissionDbEntry) new_entry = NULL;
 
@@ -457,7 +457,7 @@ handle_set_permission (XdgPermissionStore     *object,
         }
     }
 
-  permissions_id = app;
+  permissions_id = permissions_id_for_app_id (app);
   new_entry = permission_db_entry_set_app_permissions (entry, permissions_id,
                                                        (const char **) permissions);
   permission_db_set_entry (table->db, id, new_entry);
diff --git a/tests/test_permission_store.py b/tests/test_permission_store.py
index 1e6967d..b09fa3b 100644
--- a/tests/test_permission_store.py
+++ b/tests/test_permission_store.py
@@ -5,9 +5,16 @@
 import tests.xdp_utils as xdp
 
 import dbus
+import pytest
+from enum import Enum
 from gi.repository import GLib, Gio
 
 
+class PermissionsIDType(Enum):
+    PERMISSIONS_ID = 1
+    APP_ID = 2
+
+
 class PermissionStore(xdp.GDBusIface):
     def __init__(self):
         super().__init__(
@@ -91,15 +98,21 @@ class TestPermissionStore:
         )
         assert int(portal_version) == 2
 
-    def test_delete_race(self, portals, dbus_con, xdp_app_info):
+    @pytest.mark.parametrize("permission_id_type", [*PermissionsIDType])
+    def test_delete_race(self, portals, dbus_con, xdp_app_info, permission_id_type):
         permission_store_intf = PermissionStore()
         finished_count = 0
 
         table = "inhibit"
         id = "inhibit"
-        permissions_id = xdp_app_info.permissions_id
         perms = ["logout", "suspend"]
 
+        permissions_id = xdp_app_info.permissions_id
+        other_permissions_id = xdp_app_info.app_id
+        if permission_id_type == PermissionsIDType.APP_ID:
+            permissions_id = xdp_app_info.app_id
+            other_permissions_id = xdp_app_info.permissions_id
+
         def cb(_):
             nonlocal finished_count
 
@@ -112,6 +125,7 @@ class TestPermissionStore:
         permission_store_intf.DeleteAsync(table, id, cb)
 
         xdp.wait_for(lambda: finished_count >= 3)
+        finished_count = 0
 
         try:
             permission_store_intf.Lookup(table, id)
@@ -123,11 +137,15 @@ class TestPermissionStore:
         permission_store_intf.SetPermissionAsync(
             table, True, id, permissions_id, perms, cb
         )
+        permission_store_intf.SetPermissionAsync(
+            table, True, id, other_permissions_id, perms, cb
+        )
         permission_store_intf.SetPermissionAsync(table, True, id, "a", perms, cb)
         permission_store_intf.SetPermissionAsync(table, True, id, "b", perms, cb)
         permission_store_intf.DeletePermissionAsync(table, id, "a", cb)
 
-        xdp.wait_for(lambda: finished_count >= 6)
+        xdp.wait_for(lambda: finished_count >= 5)
+        finished_count = 0
 
         result, _ = permission_store_intf.Lookup(table, id)
         perms_out = result.unpack()[0]
@@ -138,13 +156,27 @@ class TestPermissionStore:
         permission_store_intf.DeletePermissionAsync(table, id, "a", cb)
         permission_store_intf.DeletePermissionAsync(table, id, permissions_id, cb)
 
-        xdp.wait_for(lambda: finished_count >= 9)
+        xdp.wait_for(lambda: finished_count >= 4)
+        finished_count = 0
 
         result, _ = permission_store_intf.Lookup(table, id)
         perms_out = result.unpack()[0]
         assert perms_out == {}
 
-    def test_change(self, portals, dbus_con, xdp_app_info):
+        permission_store_intf.SetPermissionAsync(
+            table, True, id, permissions_id, perms, cb
+        )
+        permission_store_intf.DeletePermissionAsync(table, id, other_permissions_id, cb)
+
+        xdp.wait_for(lambda: finished_count >= 2)
+        finished_count = 0
+
+        result, _ = permission_store_intf.Lookup(table, id)
+        perms_out = result.unpack()[0]
+        assert perms_out == {}
+
+    @pytest.mark.parametrize("permission_id_type", [*PermissionsIDType])
+    def test_change(self, portals, dbus_con, xdp_app_info, permission_id_type):
         permission_store_intf = PermissionStore()
         changed_count = 0
 
@@ -153,6 +185,9 @@ class TestPermissionStore:
         perms = ["one", "two"]
         permissions_id = xdp_app_info.permissions_id
 
+        if permission_id_type == PermissionsIDType.APP_ID:
+            permissions_id = xdp_app_info.app_id
+
         def cb_changed1(results):
             nonlocal changed_count
 
@@ -161,7 +196,7 @@ class TestPermissionStore:
             assert cb_table == table
             assert cb_id == id
             assert not deleted
-            assert cb_perms[permissions_id] == perms
+            assert cb_perms[xdp_app_info.permissions_id] == perms
 
             changed_count += 1
 
@@ -236,7 +271,8 @@ class TestPermissionStore:
         assert perms_out == {}
         assert data_out == data
 
-    def test_create(self, portals, dbus_con, xdp_app_info):
+    @pytest.mark.parametrize("permission_id_type", [*PermissionsIDType])
+    def test_create(self, portals, dbus_con, xdp_app_info, permission_id_type):
         permission_store_intf = PermissionStore()
 
         table = "inhibit"
@@ -244,6 +280,9 @@ class TestPermissionStore:
         permissions_id = xdp_app_info.permissions_id
         perms = ["logout", "suspend"]
 
+        if permission_id_type == PermissionsIDType.APP_ID:
+            permissions_id = xdp_app_info.app_id
+
         try:
             permission_store_intf.SetPermission(
                 table,
@@ -259,7 +298,8 @@ class TestPermissionStore:
 
         permission_store_intf.SetPermission(table, True, id, permissions_id, perms)
 
-    def test_delete(self, portals, dbus_con, xdp_app_info):
+    @pytest.mark.parametrize("permission_id_type", [*PermissionsIDType])
+    def test_delete(self, portals, dbus_con, xdp_app_info, permission_id_type):
         permission_store_intf = PermissionStore()
 
         table = "inhibit"
@@ -267,6 +307,9 @@ class TestPermissionStore:
         permissions_id = xdp_app_info.permissions_id
         perms = ["logout", "suspend"]
 
+        if permission_id_type == PermissionsIDType.APP_ID:
+            permissions_id = xdp_app_info.app_id
+
         try:
             permission_store_intf.Delete(table, id)
             assert False, "This statement should not be reached"
@@ -283,14 +326,20 @@ class TestPermissionStore:
         except GLib.GError as e:
             assert "org.freedesktop.portal.Error.NotFound" in e.message
 
-    def test_get_permission(self, portals, dbus_con, xdp_app_info):
+    @pytest.mark.parametrize("permission_id_type", [*PermissionsIDType])
+    def test_get_permission(self, portals, dbus_con, xdp_app_info, permission_id_type):
         permission_store_intf = PermissionStore()
 
         table = "notifications"
         id = "notification"
         permissions_id = xdp_app_info.permissions_id
+        other_permissions_id = xdp_app_info.app_id
         perms = ["yes"]
 
+        if permission_id_type == PermissionsIDType.APP_ID:
+            permissions_id = xdp_app_info.app_id
+            other_permissions_id = xdp_app_info.permissions_id
+
         try:
             permission_store_intf.GetPermission(table, id, permissions_id)
             assert False, "This statement should not be reached"
@@ -303,6 +352,10 @@ class TestPermissionStore:
         permissions = result.unpack()[0]
         assert permissions == perms
 
+        result, _ = permission_store_intf.GetPermission(table, id, other_permissions_id)
+        permissions = result.unpack()[0]
+        assert permissions == perms
+
         result, _ = permission_store_intf.GetPermission(table, id, "no-such-app")
         permissions = result.unpack()[0]
         assert permissions == []
